pkg/azure/validation/validation.go

```diff
@@ -0,0 +1,288 @@
+}
+
+func validateOSDisk(osDisk api.AzureOSDisk, fldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+	if utils.IsEmptyString(osDisk.CreateOption) {
+		allErrs = append(allErrs, field.Required(fldPath.Child("createOption"), "must provide createOption"))
+	}
+	if osDisk.DiskSizeGB <= 0 {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("diskSizeGB"), osDisk.DiskSizeGB, "OSDisk size must be positive and greater than 0"))
+	}
+	return allErrs
+}
+
+func validateDataDisks(disks []api.AzureDataDisk, fldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+	if disks == nil {
+		return allErrs
+	}
+
+	luns := make(map[int32]int, len(disks))
+	for _, disk := range disks {
+		if disk.Lun == nil {
+			allErrs = append(allErrs, field.Required(fldPath.Child("lun"), "must provide lun"))
+		} else {
+			if *disk.Lun < 0 {
+				allErrs = append(allErrs, field.Invalid(fldPath.Child("lun"), *disk.Lun, "lun must be a positive number"))
+			} else {
+				luns[*disk.Lun]++
+			}
+		}
+		if disk.DiskSizeGB <= 0 {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("diskSizeGB"), disk.DiskSizeGB, "DataDisk size must be positive and greater than 0"))
+		}
+		if utils.IsEmptyString(disk.StorageAccountType) {
+			allErrs = append(allErrs, field.Required(fldPath.Child("storageAccountType"), "must provide storageAccountType"))
+		}
+	}
+
+	for lun, numOccurrence := range luns {
+		if numOccurrence > 1 {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("lun"), lun, fmt.Sprintf("DataDisk Lun '%d' duplicated %d times, Lun must be unique", lun, numOccurrence)))
+		}
+	}
+
+	return allErrs
+}
+
+func validateAvailabilityAndScalingConfig(properties api.AzureVirtualMachineProperties, fldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+
+	isZoneConfigured := properties.Zone != nil
+	isAvailabilitySetConfigured := properties.AvailabilitySet != nil && !utils.IsEmptyString(properties.AvailabilitySet.ID)
+	isVirtualMachineScaleSetConfigured := properties.VirtualMachineScaleSet != nil && !utils.IsEmptyString(properties.VirtualMachineScaleSet.ID)
+
+	// check if both zone is configured and one or both of [availabilitySet, virtualMachineScaleSet] is configured
+	if !exactlyOneShouldBeTrue(isZoneConfigured, isAvailabilitySetConfigured || isVirtualMachineScaleSetConfigured) {
+		allErrs = append(allErrs, field.Forbidden(fldPath.Child("zone|.availabilitySet|.virtualMachineScaleSet"), "Either a Zone can be specified or one of (AvailabilitySet or VirtualMachineScaleSet) can be set."))
+	}
+
+	// if zone is not configured then check that only one of [availabilitySet, virtualMachineScaleSet] is configured
+	if properties.Zone == nil && !exactlyOneShouldBeTrue(isAvailabilitySetConfigured, isVirtualMachineScaleSetConfigured) {
+		allErrs = append(allErrs, field.Forbidden(fldPath.Child("availabilitySet|.virtualMachineScaleSet"), "Must only configure an AvailabilitySet or VirtualMachineScaleSet but not both."))
+	}
+
```

Comment:
these essentially mean that exactly one of `zone`, `AvailabilitySet` and `VirtualMachineScaleSet` should be configured, right? Then why not do it in just one `exactlyOneShouldBeTrue()`

Also what is the relation b/w `zone` with the other parameters specified above, that we are having such a check? As I don't see any such relation mentioned in the API comments

Reply:
Well in the API the official documentation only states that if a VMSS and AvailabilitySet cannot be set together. I did not find a mention of zone. However the old code has that additional restriction. So yes you are right that we could have just used `exactlyOneShouldBeTrue` for all 3 parameters but i would like to now look into why this additional validation for zone is in place as the API does not mention that.

Reply:
From the offician MS Azure documentation, availability set should be used in regions that do not support availability zones. (https://learn.microsoft.com/en-us/azure/virtual-machines/availability-set-overview)
From the docs what i can understand is that availability set offers higher availability by spreading the VM's across fault domains in a single location. Therefore specifying zones along with AvailabilitySet would ideally make not much sense. I had tried to create an AvailabilitySet but the user does not have sufficient permissions to create one.

Reply:
If you look at https://learn.microsoft.com/en-us/azure/virtual-machine-scale-sets/virtual-machine-scale-sets-orchestration-modes?WT.mc_id=modinfra-11089-socuff#high-availability then it clearly states that AvailabilitySet does not support multiple zones, but it should still support a single zone, right? `VirtualMachine.Zones` is a `[]*string` and therefore also allows specifying one zone. Will this work then is something that we should ideally test.
For VMSS it supports both single and multi-zone but then that configuration should be done when creating VMSS. When we create a VM we only give a reference to either of VMSS or AvailabilitySet.

Reply:
thanks for taking such a deeper look. as decided, we will keep `exactlyOneShouldBeTrue` for the three fields

Reply:
done the change and adjusted the test.

Reply:
thanks. could you also add a comment to mention why we are only checking `exactlyOneShouldBeTrue`. Basically a comment specifying setting `zone` means MCM will manage the machine, and `availabilitySet|VirtualMachineScaleSet` would mean they'll manage.

Reply:
This is done.